Support running Temporal tests in Node 19 (cont'd for staging tests)

justingrant opened the pull request:
Following up from #3750, this PR supports running Temporal "staging" tests (https://github.com/tc39/test262/tree/main/test/staging) under Node 19.

For context, `Intl.DateTimeFormat` implementations in Node 19 have ICU format changes:
- replace some spaces with non-breaking-space unicode characters
- change German date separators (these may be bugs)
- months in German are shown with a leading zero, but only in Year/Month formats (also may be a bug)

This PR makes affected tests more resilient to `Intl.DateTimeFormat` output changes between Node versions. They'll continue to work even if the behavior (e.g. in German) turns out to be a bug and is fixed in the platform.

gibson042 commented on test/staging/Intl402/Temporal/old/date-time-format.js:
```diff
@@ -616,42 +639,56 @@ assert.deepEqual(at.formatToParts(end), [
+  us.formatRange(date2.withCalendar(usCalendar).toPlainMonthDay(), date1.withCalendar(usCalendar).toPlainMonthDay()),
+  `2/20${dateRangeSeparator}11/18`
+);
+// In Node 19.2.0, apparently the german Month/Day format uses regular spaces, but all other formats and locales seem to use thin spaces ('\u{2009}');
```
There seems to be a lot of variability with de-AT; perhaps it would make sense to pick a different non-English locale if these comparisons truly are important?

justingrant replied:
After quite a bit of digging, the problem looks like a regression in CLDR 42's German date formats. Frustratingly, the Node 19 build used by GitHub Actions picks up ICU 72 (which includes CLDR 42 with the problems) but `brew install node` on my local machine currently includes ICU 71/CLDR 41. This explains why I only saw it in CI and why this problem has been so frustrating to track down!

I filed a bug in CLDR's JIRA to fix the formats: https://unicode-org.atlassian.net/browse/CLDR-16243

Now that I understand the actual root cause, I can remove a lot of the German-specific code because only month/year formats are affected, not all date formats. This should clean up this PR quite a bit. I'll do that tomorrow and will ping you to review again once that's done.

Thanks for pushing me to dig into this!

justingrant replied:
I ended up leaving German as the locale because AFAICT would be a lot of work to change it throughout all these tests, and it's relatively straightforward to write tests that ignore the German issues. I added TODO comments that encourage future contributors to consider tests that have less dependence on CLDR formats.

justingrant commented on test/staging/Intl402/Temporal/old/date-time-format.js:
```diff
@@ -9,7 +9,23 @@ includes: [deepEqual.js]
+// than other date formats.
+const germanMonthYearSeparator = new Intl.DateTimeFormat("de-AT", { year: "numeric", month: "numeric" }).formatToParts(
+  0
+)[1].value;
```
Good questions!  I think your changes are good ones, I'll include them in the next push of this PR, hopefully tomorrow.  Also (see discussion below about the German formats) it's likely that much of the German-specific code in this PR will go away, which should further simplify things and shrink the diff.

I agree that it would be good to further reduce the dependency of these tests on CLDR string formats. But I don't think I'll have time to do that work myself. For this PR, I was kinda hoping to do the minimum changes required to get Test262 to pass in both Node 18 and Node 19 so we don't block anyone's testing as folks start upgrading their CIs to v19.

If OK with you, I'd like to leave the "refactor to reduce CLDR format dependency" as an exercise for the future. Perhaps whoever ports these "staging" tests into the mainstream Test262 could also look into refactoring them? Does this sound OK?

justingrant replied:
Yep, done!

justingrant commented:
@gibson042 - I made the changes you recommended, rebased to main, and cleaned up commits by running prettier first and then layering my changes on top so the diff will be easier to review.

gibson042 commented on test/staging/Intl402/Temporal/old/date-time-format.js:
```diff
@@ -518,6 +537,15 @@ assert.deepEqual(us.formatToParts(t.withCalendar(usCalendar).toPlainYearMonth())
     value: "2020"
   }
 ]);
+
+// Work around variability in month format caused by https:// unicode-org.atlassian.net/browse/CLDR-16243
+var deMonthInYearMonthHasLeadingZero = new Intl.DateTimeFormat("de-AT", {
+  year: "numeric",
+  month: "numeric"
+})
+  .formatToParts(new Date(2000, 3, 1))
+  .find((part) => part.type === "month")
+  .value.startsWith("0");
```
This appears to be unused.
```suggestion
```

justingrant replied:
Whoops, looks like I lost this check during the last rebase. The next line should have used this value in a conditional:
```js
assert.deepEqual(at.formatToParts(t.withCalendar(atCalendar).toPlainYearMonth()), [
  {
    type: "month",
    value: deMonthInYearMonthHasLeadingZero ? "02" : "2"
  },
```

Will be fixed in the next push. 

It's actually lucky you found this, because it looks like something went wrong on the proposal-temporal side of this PR and it wasn't picking up the right Test262 subproject commit. Otherwise the proposal-temporal PR should have failed. Investigating that now.  Thanks!

gibson042 commented on test/staging/Intl402/Temporal/old/date-time-format.js:
```diff
@@ -680,10 +714,11 @@ assert.sameValue(
+assert.sameValue(at.formatRange(start, end), `30.12.1922${deDateRangeSeparator}26.12.1991`);
 
-// should throw a TypeError when called with dissimilar types", () => assert.throws(TypeError, () => us.formatRange(Temporal.Instant.from(t1), Temporal.PlainDateTime.from(t2))));
+// should throw a TypeError when called with dissimilar types",
+//  () => assert.throws(TypeError, () => us.formatRange(Temporal.Instant.from(t1), Temporal.PlainDateTime.from(t2))));
```
Looks like this is actually tested below.
```suggestion
```

justingrant replied:
Looks like the original port from demitasse tests forgot to remove this commented line. I'll remove it.

gibson042 commented on test/staging/Intl402/Temporal/old/instant-toLocaleString.js:
```diff
@@ -7,6 +7,12 @@ description: Instant.toLocaleString()
 ---*/
 
+// Tolerate implementation variance by expecting consistency without being prescriptive.
+const usDayPeriodSpace =
+  new Intl.DateTimeFormat("en-US", { timeStyle: "short" })
```
Shouldn't this be used in the first assertion, or are we looking at a difference between "en-US" and "en"?
```diff
-assert.sameValue(`${instant.toLocaleString("en", { timeZone: "America/New_York" })}`, "11/18/1976, 9:23:30 AM");
+assert.sameValue(`${instant.toLocaleString("en", { timeZone: "America/New_York" })}`, `11/18/1976, 9:23:30${usDayPeriodSpace}AM`);
```

justingrant replied:
Good catch. It looks like there's a mix in these tests between "en" & "de" vs. "en-US" and "de-AT", which I assume was an oversight from whoever originally wrote these tests. I'll just normalize them all into "en-US" and "de-AT" throughout all these files, and I'll put that into a separate commit to keep the noise down.

gibson042 commented on test/staging/Intl402/Temporal/old/time-toLocaleString.js:
```diff
@@ -7,6 +7,12 @@ description: time.toLocaleString()
 ---*/
 
+// Tolerate implementation variance by expecting consistency without being prescriptive.
+const usDayPeriodSpace =
+  new Intl.DateTimeFormat("en-US", { timeStyle: "short" })
```
Shouldn't this be used in the assertions, or are we looking at a difference between "en-US" and "en"?

justingrant commented:
@gibson042 All code review issues should be resolved, but there's been enough change that it may be worth one more review before merging.

I apologize for all the commit notification spam from this PR. I wasn't able to figure out how to get my local install of Node 19 to use ICU 72 like GH Actions does, so I had to use CI to validate all changes. Grrrrr.
